controllers/routing/routing_controller.go

```diff
@@ -66,11 +68,20 @@ func (r *PlatformRoutingReconciler) Reconcile(ctx context.Context, req ctrl.Requ
 		return ctrl.Result{}, fmt.Errorf("failed getting resource: %w", err)
 	}
 
-	// TODO(mvp) if !sourceRes.GetDeletionTimestamp().IsZero() { handle removal of dependant resources }
+	if !sourceRes.GetDeletionTimestamp().IsZero() {
+		return r.HandleResourceDeletion(ctx, sourceRes)
+	}
 
 	r.log.Info("triggered routing reconcile", "namespace", req.Namespace, "name", req.Name)
 
 	var errs []error
+
+	if controllerutil.AddFinalizer(sourceRes, metadata.Finalizers.Routing) {
+		if errUpdate := r.Update(ctx, sourceRes); errUpdate != nil {
```

Comment:
I think we should return here instead of accumulating error. Otherwise we might fail updating watched resource with added finalizer and continue creating potentially orphaned routing resources. So we're still facing the original problem.

Reply:
[50cb72b](https://github.com/maistra/odh-platform/pull/39/commits/50cb72b69dbc56c67a394eefc709c9131e5054af)